lib/src/matrix/decomposition/singular_value.dart

```diff
@@ -1,483 +1,951 @@
+          }
+
+          result.setUnchecked(j, k, value);
+        }
+      }
+
+      return result;
+    }
+    if (input is Vector<double>) {
+      // Ax=b where A is an m x n matrix
+      // Check that b is a column vector with m entries
+      if (_u.rowCount != input.count) {
+        throw ArgumentError('All vectors must have the same dimensionality.');
+      }
+
+      final result = Vector<double>(DataType.float64, _vt.colCount);
+
+      var mn = min(_u.rowCount, _vt.colCount);
+      var tmp = List.filled(_vt.colCount, 0.0);
+      double value;
+      for (var j = 0; j < _vt.colCount; j++) {
+        value = 0;
+        if (j < mn) {
+          for (var i = 0; i < _u.rowCount; i++) {
+            value += _u.getUnchecked(i, j) * input[i];
+          }
+
+          value /= _s[j];
+        }
+
+        tmp[j] = value;
+      }
+
+      for (var j = 0; j < _vt.colCount; j++) {
+        value = 0;
+        for (var i = 0; i < _vt.colCount; i++) {
+          value += _vt.getUnchecked(i, j) * tmp[i];
+        }
+
+        result[j] = value;
+      }
+
+      return result;
     }
-    return result;
   }
 
-  /// Return the two norm.
-  double get norm2 => _s[0];
+  static List<double> _columnMajorValuesOf(Matrix<num> matrix) {
+    final list = List.filled(matrix.rowCount * matrix.colCount, 0.0);
+    var index = 0;
+    for (var col = 0; col < matrix.colCount; col++) {
+      for (var row = 0; row < matrix.rowCount; row++) {
+        list[index++] = matrix.getUnchecked(row, col).toDouble();
+      }
+    }
+    return list;
+  }
 
-  /// Return the two norm condition number.
-  double get cond => _s[0] / _s[math.min(_m, _n) - 1];
+  /// Given the Cartesian coordinates (da, db) of a point p, these function return the parameters da, db, c, and s
+  /// associated with the Givens rotation that zeros the y-coordinate of the point.
+  ///
+  /// Note that Dart is a pass-by-value language, so modifiy the code little bit.
+  static List<double> _rotg(double da, double db) {
+    double c, s; // out
+
+    final absda = da.abs();
+    final absdb = db.abs();
+    final roe = (absda > absdb) ? da : db;
+    final scale = absda + absdb;
+
+    double r, z;
+    if (scale == 0.0) {
+      c = 1.0;
+      s = 0.0;
+      r = 0.0;
+      z = 0.0;
+    } else {
+      final sda = da / scale;
+      final sdb = db / scale;
+      r = scale * sqrt((sda * sda) + (sdb * sdb));
+      if (roe < 0.0) {
+        r = -r;
+      }
 
-  /// Return the effective numerical matrix rank.
-  int get rank {
-    final eps = math.pow(2.0, -52.0);
-    final tol = math.max(_m, _n) * s[0] * eps;
-    var r = 0;
-    for (var i = 0; i < _s.length; i++) {
-      if (_s[i] > tol) {
-        r++;
+      c = da / r;
+      s = db / r;
+      z = 1.0;
+      if (absda > absdb) {
+        z = s;
+      }
+
+      if (absdb >= absda && c != 0.0) {
+        z = 1.0 / c;
       }
     }
-    return r;
+
+    da = r;
+    db = z;
+
+    return [da, db, c, s];
```

Comment:
Fine for now, but should replace with a [record type](https://github.com/dart-lang/language/blob/master/accepted/future-releases/records/records-feature-specification.md) of Dart 3.0 (maybe add a TODO).

Reply:
Added TODO